### PR TITLE
Finding No. 6.1.4 Add a content security policy header

### PR DESCRIPTION
**Pull request by mattonfoot**
# ITHC Finding 6.1.4 - Add Content Security Policy

## Recommendation
Implement the missing security headers highlighted as missing in the “Details” section, above. Further information on how to achieve this is provided in the “References” section, below.It is further noted that certain HTTP security headers have been known to highlight usability issues in application functionality. When implementing security headers for the first time, rigorous testing is recommended before deploying changes to live environments.

## Mitigation
Add host “content-security-policy” header to rails

### CODE REVIEW

File: config/initializers/secure_headers.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    form_action: %w['self'],
+    frame_ancestors: %w['self'],
+    frame_src: %w['self'],
+    img_src: %W['self' *.gov.uk] + google_analytics,
+    manifest_src: %w['self'],
+    media_src: %w['self'],
+    script_src: %W['self' 'unsafe-inline' 'unsafe-eval' *.gov.uk] + google_analytics,
+    style_src: %w['self' 'unsafe-inline' *.gov.uk] + google_analytics,
+    worker_src: %w['self'],
+    upgrade_insecure_requests: !Rails.env.development?, # see https://www.w3.org/TR/upgrade-insecure-requests/
```
Comment (kicferk1) [inline config/initializers/secure_headers.rb:30]: Do we need this? Pretty sure our server doesn't accept insecure requests to begin with. 
If we do need this, do we think this condition should be `!(Rails.env.development? || Rails.env.test?)`?

File: config/initializers/secure_headers.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  config.x_content_type_options = "nosniff"
+  config.x_xss_protection = "1; mode=block"
+  config.x_download_options = "noopen"
+  config.x_permitted_cross_domain_policies = "none"
+  config.referrer_policy = %w[origin-when-cross-origin strict-origin-when-cross-origin]
+
+  google_analytics = %w[www.google-analytics.com ssl.google-analytics.com *.googletagmanager.com tagmanager.google.com *.googleusercontent.com *.gstatic.com]
+
+  config.csp = {
+    default_src: %w['none'],
```
Comment (kicferk1) [inline config/initializers/secure_headers.rb:15]: It's a shame that rubocop thinks that we shouldn't have strings that include `'` characters.

**Comment by kicferk1** [comment]
Also, when that goes in, I will steal it for E&L :)

File: config/initializers/secure_headers.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
```
Comment (ethanmills) [inline config/initializers/secure_headers.rb:1]: I think we should either use `config/initializers/content_security_policy.rb` or remove it

File: config/initializers/secure_headers.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    frame_ancestors: %w['self'],
+    frame_src: %w['self'],
+    img_src: %W['self' *.gov.uk] + google_analytics,
+    manifest_src: %w['self'],
+    media_src: %w['self'],
+    script_src: %W['self' 'unsafe-inline' 'unsafe-eval' *.gov.uk] + google_analytics,
+    style_src: %w['self' 'unsafe-inline' *.gov.uk] + google_analytics,
+    worker_src: %w['self'],
+    upgrade_insecure_requests: !Rails.env.development?, # see https://www.w3.org/TR/upgrade-insecure-requests/
+    report_uri: %w[/csp_reports],
```
Comment (ethanmills) [inline config/initializers/secure_headers.rb:31]: We should consider using the reporting endpoint from Sentry

Comment (mattonfoot) [reply config/initializers/secure_headers.rb:31]: Added an internal route with logging to console with notifications. Or should I just log direct from controller

File: config/initializers/secure_headers.rb (hunk 3 above)
Comment (mattonfoot) [reply config/initializers/secure_headers.rb:1]: As it is all commented out and it is a ?default rails file? I have left it in, I found on SE that these files were best kept in when you do an upgrade to rails

File: config/initializers/secure_headers.rb (hunk 2 above)
Comment (mattonfoot) [reply config/initializers/secure_headers.rb:15]: Rubocop didn't pick anything up for me, have I missed something I should change, happy to restyle even without rubocop errors

File: config/initializers/secure_headers.rb (hunk 1 above)
Comment (mattonfoot) [reply config/initializers/secure_headers.rb]: have removed that line, because your correct it was in just so you can run it locally on http

**Review by jamesrose — APPROVED** [review]

**Pull request merged**